### ISSUE TITLE
regex (?{}) setting flag variable gets lost upon leaving

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13426]
Migrated from [rt.perl.org#120554](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120554) (status was 'resolved')

Searchable as RT120554$

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

Looks like when the flag is set it works properly for the first time the regex is executed\, but then doesn't get set second time through\.  Possibly referencing wrong object\.

Versions tried on&#8203;:
This is perl 5\, version 16\, subversion 2 \(v5\.16\.2\) built for i686\-linux
This is perl 5\, version 14\, subversion 2 \(v5\.14\.2\) built for cygwin\-thread\-multi\-64int

Test code can be found here&#8203;:
http&#8203;://ideone\.com/gBvio0

By changing the value being set on the line&#8203;: IF YOU CHANGE THIS VALUE\, IT WILL STAY THAT VALUE THE 2ND TIME THROUGH\, it will not be changed upon leaving the regex\.

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

On Sat Nov 16 16&#8203;:07&#8203;:09 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> Looks like when the flag is set it works properly for the first time
> the regex is executed\, but then doesn't get set second time through\.
> Possibly referencing wrong object\.
> 
> Versions tried on&#8203;:
> This is perl 5\, version 16\, subversion 2 \(v5\.16\.2\) built for i686\-
> linux
> This is perl 5\, version 14\, subversion 2 \(v5\.14\.2\) built for cygwin\-
> thread\-multi\-64int
> 
> Test code can be found here&#8203;:
> http&#8203;://ideone\.com/gBvio0
> 
> By changing the value being set on the line&#8203;: IF YOU CHANGE THIS VALUE\,
> IT WILL STAY THAT VALUE THE 2ND TIME THROUGH\, it will not be changed
> upon leaving the regex\.


This appears to be a scoping issue\.  If the variable $complete is declared in the base\(?\) scope \(i\.e\. it's not in a block\) then it works every time\.  But by declaring it inside a block\, it fails to get the correct scope the 2nd time\.

This can be seen in test code here&#8203;:
http&#8203;://ideone\.com/PhfilX

This is definitely a bug\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sat Nov 16 16&#8203;:07&#8203;:09 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> Looks like when the flag is set it works properly for the first time
> the regex is executed\, but then doesn't get set second time through\.
> Possibly referencing wrong object\.
> 
> Versions tried on&#8203;:
> This is perl 5\, version 16\, subversion 2 \(v5\.16\.2\) built for i686\-
> linux
> This is perl 5\, version 14\, subversion 2 \(v5\.14\.2\) built for cygwin\-
> thread\-multi\-64int
> 
> Test code can be found here&#8203;:
> http&#8203;://ideone\.com/gBvio0
> 
> By changing the value being set on the line&#8203;: IF YOU CHANGE THIS VALUE\,
> IT WILL STAY THAT VALUE THE 2ND TIME THROUGH\, it will not be changed
> upon leaving the regex\.


This appears to be a scoping issue\.  If the variable $complete is declared in the base\(?\) scope \(i\.e\. it's not in a block\) then it works every time\.  But by declaring it inside a block\, it fails to get the correct scope the 2nd time\.

This can be seen in test code here&#8203;:
http&#8203;://ideone\.com/PhfilX

This is definitely a bug\.

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

On Sat Nov 16 16&#8203;:31&#8203;:24 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> On Sat Nov 16 16&#8203;:07&#8203;:09 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> > Looks like when the flag is set it works properly for the first time
> > the regex is executed\, but then doesn't get set second time through\.
> > Possibly referencing wrong object\.
> >
> > Versions tried on&#8203;:
> > This is perl 5\, version 16\, subversion 2 \(v5\.16\.2\) built for i686\-
> > linux
> > This is perl 5\, version 14\, subversion 2 \(v5\.14\.2\) built for cygwin\-
> > thread\-multi\-64int
> >
> > Test code can be found here&#8203;:
> > http&#8203;://ideone\.com/gBvio0
> >
> > By changing the value being set on the line&#8203;: IF YOU CHANGE THIS
> > VALUE\,
> > IT WILL STAY THAT VALUE THE 2ND TIME THROUGH\, it will not be changed
> > upon leaving the regex\.
> 
> 
> This appears to be a scoping issue\.  If the variable $complete is
> declared in the base\(?\) scope \(i\.e\. it's not in a block\) then it works
> every time\.  But by declaring it inside a block\, it fails to get the
> correct scope the 2nd time\.
> 
> This can be seen in test code here&#8203;:
> http&#8203;://ideone\.com/PhfilX
> 
> This is definitely a bug\.

The following code is a much condensed example of the failure&#8203;:
&nbsp;  http&#8203;://ideone\.com/X4j88u

&nbsp;    \#\!/usr/bin/perl
&nbsp;    
&nbsp;    use strict;
&nbsp;    use warnings;
&nbsp;    my $x = 1;
&nbsp;    while \($x\)
&nbsp;    \{
&nbsp;        my $complete = 2;
&nbsp;        if \($complete\)
&nbsp;        \{
&nbsp;            use re 'eval';
&nbsp;            "" =~ /\(?\{$complete\-=1\}\)/;
&nbsp;            print "COMPLETE = $complete\\n";
&nbsp;        \}
&nbsp;        $x = $complete;
&nbsp;    \}

Moving the declaration of $complete to the base scope allow termination of the loop\.
&nbsp;  http&#8203;://ideone\.com/WEtHNv

&nbsp;    \#\!/usr/bin/perl
&nbsp;    
&nbsp;    use strict;
&nbsp;    use warnings;
&nbsp;    my $x = 1;
&nbsp;    my $complete = 2;
&nbsp;    while \($x\)
&nbsp;    \{
&nbsp;        if \($complete\)
&nbsp;        \{
&nbsp;            use re 'eval';
&nbsp;            "" =~ /\(?\{$complete\-=1\}\)/;
&nbsp;            print "COMPLETE = $complete\\n";
&nbsp;        \}
&nbsp;        $x = $complete;
&nbsp;    \}


A

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Nov 16 17&#8203;:19&#8203;:47 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:


> >
> > This is definitely a bug\.
> 
> The following code is a much condensed example of the failure&#8203;:
>   http&#8203;://ideone\.com/X4j88u
> 
> \#\!/usr/bin/perl
> 
> use strict;
> use warnings;
> my $x = 1;
> while \($x\)
> \{
>         my $complete = 2;
>         if \($complete\)
>         \{
>                 use re 'eval';
>                 "" =~ /\(?\{$complete\-=1\}\)/;
>                 print "COMPLETE = $complete\\n";
>         \}
>         $x = $complete;
> \}
> 

I don't see any bug in Perl here\.  When 'my $complete' is declared within the loop it is a declaration of a fresh variable\.  You can see this if you insert a line like&#8203;:

&nbsp;   print "At the top of the while loop\\n";

just after the 'while' and run the code through the Perl debugger\.  When you reach "At the top of the while loop"\, you will see that $complete is undef\.  You then assign '2' to a freshly declared $complete\.  At the bottom of the 'while' loop\, you assign '1' to '$x'\, causing an infinite loop\.

Your use of "use re 'eval'" and ?\{\} is irrelevant to the problem\.  If you were to simply say&#8203;:

&nbsp;   $complete \-= 1;

\.\.\. instead of using ?\{\}\, but still declared 'my $complete inside the 'while' loop\, you would still get an infinite loop\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

On Sat Nov 16 18&#8203;:47&#8203;:33 2013\, jkeenan wrote&#8203;:
> On Sat Nov 16 17&#8203;:19&#8203;:47 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> 
> 
> > >
> > > This is definitely a bug\.
> >
> > The following code is a much condensed example of the failure&#8203;:
> >   http&#8203;://ideone\.com/X4j88u
> >
> > \#\!/usr/bin/perl
> >
> > use strict;
> > use warnings;
> > my $x = 1;
> > while \($x\)
> > \{
> >         my $complete = 2;
> >         if \($complete\)
> >         \{
> >                 use re 'eval';
> >                 "" =~ /\(?\{$complete\-=1\}\)/;
> >                 print "COMPLETE = $complete\\n";
> >         \}
> >         $x = $complete;
> > \}
> >
> 
> I don't see any bug in Perl here\.  When 'my $complete' is declared
> within the loop it is a declaration of a fresh variable\.  You can see
> this if you insert a line like&#8203;:
> 
> print "At the top of the while loop\\n";
> 
> just after the 'while' and run the code through the Perl debugger\.
> When you reach "At the top of the while loop"\, you will see that
> $complete is undef\.  You then assign '2' to a freshly declared
> $complete\.  At the bottom of the 'while' loop\, you assign '1' to '$x'\,
> causing an infinite loop\.
> 
> Your use of "use re 'eval'" and ?\{\} is irrelevant to the problem\.  If
> you were to simply say&#8203;:
> 
> $complete \-= 1;
> 
> \.\.\. instead of using ?\{\}\, but still declared 'my $complete inside the
> 'while' loop\, you would still get an infinite loop\.

I was looking further into this and my example was poorly written\.  I should have at least stated what the expected output was\, which should have been&#8203;:

&nbsp;  COMPLETE = 1
&nbsp;  COMPLETE = 1
&nbsp;  COMPLETE = 1
&nbsp;  \.\.\.

What actually was output was&#8203;:

&nbsp;  COMPLETE = 1
&nbsp;  COMPLETE = 2
&nbsp;  COMPLETE = 2
&nbsp;  \.\.\.

Here is a better example&#8203;:

\#\!/usr/bin/perl
\# Expected output&#8203;:
\#   3\) COMPLETE = 1
\#   2\) COMPLETE = 1
\#   1\) COMPLETE = 1
\# Actual output using my inside the loop&#8203;:
\#   3\) COMPLETE = 1
\#   2\) COMPLETE = 2
\#   1\) COMPLETE = 2
use strict;
use warnings;
use re 'eval';
my $x = 3;

while \($x\)
\{
&nbsp;    \# using my inside loop just doesn't work
&nbsp;    my $complete;
&nbsp;    $complete = 2;
&nbsp;    "" =~ /\(?\{\-\-$complete\}\)/;
&nbsp;    \#\-\-$complete;
&nbsp;    print "$x\) COMPLETE = $complete\\n";
&nbsp;    \-\-$x;
\}

However the behaviour does make sense after a fashion\.  The regex is compiled only once\, so \(without looking at the internals\) I'm assuming that $complete is compiled as some sort of reference to the object holding the information\.  So when the new object for $complete is created\, it doesn't refer to the one referred to by the regex\.

Doing the same thing from within a function causes one of two warnings to be generated\.  If the while loop is wrapped in an anonymous function\, it spits out 'Variable "$complete" is not available at \(re\_eval 1\) line 2\.'\,  if wrapped in a static\(?\) function\, it spits out 'Variable "$complete" will not stay shared at \(re\_eval 1\) line 2\.'  These warnings seem to confirm my hypothesis\.  Also\, using a state variable instead of a my also resolves these issues\.

However\, I think that there are 2 problems here&#8203;:

1\. The above example doesn't generate any warnings\.
2\. This really shouldn't be a warning but an error to prevent hard to find bugs\.

One other thing\.  I'm not sure where it's getting the line number from in the warnings I stated\, but it doesn't change no matter what line it is on\.  I'm assuming that it has to do with the line from within the regex\.  But wouldn't that be line 1 then?

Thanks\,


Adrian Hawryluk

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

So when a ticket is rejected\, is it rejected for life?  Or can it be reopened?  Or should I submit another ticket?

Thanks\,


Adrian Hawryluk

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'rejected' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Nov 17\, 2013 at 08&#8203;:56&#8203;:40PM \-0800\, Adrian via RT wrote&#8203;:
> So when a ticket is rejected\, is it rejected for life?  Or can it be
> reopened?  Or should I submit another ticket?

Any follow\-up can still be applied to the old ticket\. The 'rejected'
status only affects searching for still\-open tickets\.

Having said that\, the ticket should remain closed; this \*was\* a bug in
perl\, but was fixed in 5\.18\.0\, when the \(?\{\}\) implementation had a major
overhaul\.

&nbsp;    $ perl5163 ~/tmp/p
&nbsp;    3\) COMPLETE = 1
&nbsp;    2\) COMPLETE = 2
&nbsp;    1\) COMPLETE = 2

&nbsp;    $ perl5180 ~/tmp/p
&nbsp;    3\) COMPLETE = 1
&nbsp;    2\) COMPLETE = 1
&nbsp;    1\) COMPLETE = 1

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

On Mon Nov 18 04&#8203;:07&#8203;:14 2013\, davem wrote&#8203;:
> Having said that\, the ticket should remain closed; this \*was\* a bug in
> perl\, but was fixed in 5\.18\.0\, when the \(?\{\}\) implementation had a major
> overhaul\.
> 
>     $ perl5163 ~/tmp/p
>     3\) COMPLETE = 1
>     2\) COMPLETE = 2
>     1\) COMPLETE = 2
> 
>     $ perl5180 ~/tmp/p
>     3\) COMPLETE = 1
>     2\) COMPLETE = 1
>     1\) COMPLETE = 1
> 

Great to know\!  Thanks\!


A

**Comment by p5pRT** [comment]
### From adrianh.bsc@gmail.com

On Wed Nov 20 03&#8203;:41&#8203;:54 2013\, adrianh\.bsc@&#8203;gmail\.com wrote&#8203;:
> On Mon Nov 18 04&#8203;:07&#8203;:14 2013\, davem wrote&#8203;:
> > Having said that\, the ticket should remain closed; this \*was\* a bug in
> > perl\, but was fixed in 5\.18\.0\, when the \(?\{\}\) implementation had a major
> > overhaul\.
> > 
> >     $ perl5163 ~/tmp/p
> >     3\) COMPLETE = 1
> >     2\) COMPLETE = 2
> >     1\) COMPLETE = 2
> > 
> >     $ perl5180 ~/tmp/p
> >     3\) COMPLETE = 1
> >     2\) COMPLETE = 1
> >     1\) COMPLETE = 1
> > 
> 
> Great to know\!  Thanks\!
> 
> 
> A

Oh\, sorry\, I don't have access to 5\.18\.0 at the moment\.  So is this using a global/state $complete or a scoped one?  Was I correct in how it compiles the \(?\{\}\) expression?

Thanks\,


A

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Nov 20\, 2013 at 03&#8203;:48&#8203;:56AM \-0800\, Adrian via RT wrote&#8203;:
> Oh\, sorry\, I don't have access to 5\.18\.0 at the moment\.  So is this
> using a global/state $complete or a scoped one?  Was I correct in how it
> compiles the \(?\{\}\) expression?

This is with $complete being a lexical var\.

Prior to 5\.18\.0\, \(?\{\}\) was mostly broken with respect to lexical vars;
at best you usually got the wrong value; at worst\, segfaults\.

The behaviour from 5\.18\.0 onwards as regards scope and closures\, is that

&nbsp;    /A\(?\{B\}\)C/

is compiled as if it were written

&nbsp;    /A/ && do\{B\} && /C/

and

&nbsp;    qr/A\(?\{B\}\)C/

is compiled as if it were written

&nbsp;    sub \{ /A/ && do\{B\} && /C/ \}


\-\- 
That he said that that that that is is is debatable\, is debatable\.
